### PR TITLE
[BEAM-2558] Migrate checkCombineFn in TestUtils to CombineFnTester

### PR DESCRIPTION
**Pull request by tgroh**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
This makes CombineFnTester significantly more discoverable, and usable
without having dependencies on the test JAR.

Update existing tests.

This is purely a move of code with minor renaming, visibility limiting, and
updates to include empty accumulators on both ends of `mergeAccumulators`.

Worth considering is also ensuring that the output (as returned by `extractOutput`)
is not the same as any of the merged accumulators (as doing so potentially interacts
poorly with a lifted Combine)

### CODE REVIEW

**Comment by tgroh** [comment]
R: @robertwb @kennknowles (either)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    checkCombineFnShardsMultipleOrders(fn, shardExponentially(input, 2), matcher);
+    checkCombineFnShardsMultipleOrders(fn, shardExponentially(input, Math.E), matcher);
+  }
+
+  private static <InputT, AccumT, OutputT> void checkCombineFnShardsMultipleOrders(
+      CombineFn<InputT, AccumT, OutputT> fn,
+      List<? extends Iterable<InputT>> shards,
+      Matcher<? super OutputT> matcher) {
+    checkCombineFnShardsWithEmptyAccumulators(fn, shards, matcher);
+    Collections.shuffle(shards); checkCombineFnShardsWithEmptyAccumulators(fn, shards, matcher);
```
Comment (robertwb) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java]: Split into two lines?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  private static <InputT, AccumT, OutputT> void checkCombineFnShardsWithEmptyAccumulators(
```
Comment (robertwb) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java]: Should we also check without empty accumulators?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    checkCombineFnShardsMultipleOrders(fn, shardEvenly(input, 2), matcher);
+    if (size > 4) {
+      checkCombineFnShardsMultipleOrders(fn, shardEvenly(input, size / 2), matcher);
+      checkCombineFnShardsMultipleOrders(
+          fn, shardEvenly(input, (int) (size / Math.sqrt(size))), matcher);
+    }
+    checkCombineFnShardsMultipleOrders(fn, shardExponentially(input, 1.4), matcher);
+    checkCombineFnShardsMultipleOrders(fn, shardExponentially(input, 2), matcher);
+    checkCombineFnShardsMultipleOrders(fn, shardExponentially(input, Math.E), matcher);
+  }
```
Comment (robertwb) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java]: This utility was written before streaming. Should we also test that style of combining, e.g.

extractOutput(
  merge(createAccumulator().addInput(v1),
    merge(createAccumulator().addInput(v2),
        merge(createAccumulator().addInput(v3),
          ...

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CombineFnTester.java]: Done.

**Comment by tgroh** [comment]
Added tests.

**Comment by XuMingmin** [comment]
retest this please

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+import org.junit.runners.JUnit4;
+
+/**
+ * Tests for {@link CombineFnTester}.
+ */
+@RunWith(JUnit4.class)
+public class CombineFnTesterTest {
+  @Test
+  public void checksMergeWithEmptyAccumulators() {
+    final AtomicBoolean sawEmpty = new AtomicBoolean(false);
```
Comment (robertwb) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java]: Why does this need to be atomic?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+        };
+
+    CombineFnTester.testCombineFn(combineFn, Arrays.asList(1, 2, 3, 4, 5), 15);
+    assertThat(sawSingleShard.get(), is(true));
+  }
+
+  @Test
+  public void checksWithShards() {
+    final AtomicBoolean sawManyShards = new AtomicBoolean();
+    CombineFn<Integer, int[], Integer> combineFn =
```
Comment (robertwb) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java]: Might be cleaner to just use an Integer for the accumulator as performance doesn't matter here.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    CombineFn<Integer, List<Integer>, Integer> combineFn =
+        new CombineFn<Integer, List<Integer>, Integer>() {
+          @Override
+          public List<Integer> createAccumulator() {
+            return new ArrayList<>();
+          }
+
+          @Override
+          public List<Integer> addInput(List<Integer> accumulator, Integer input) {
+            if (!accumulator.isEmpty() && accumulator.get(accumulator.size() - 1) > input) {
```
Comment (robertwb) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java]: This line is a bit obscure, comment?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+          public void describeTo(Description description) {}
+
+          @Override
+          protected boolean matchesSafely(Integer item) {
+            matcherUsed.set(true);
+            return item == 30;
+          }
+        };
+    CombineFnTester.testCombineFn(
+        Sum.ofIntegers(), Arrays.asList(1, 1, 2, 2, 3, 3, 4, 4, 5, 5), matcher);
```
Comment (robertwb) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java]: For this kind of test I think it's more succinct to use a matcher that matches nothing (or the wrong thing) and verify that CombineFnTester.testCombine fails. (For writing TesterTesters, my preference is generally to write a Fn that fails (just) the property you are trying to validate, and verify that the tester rejects it. However, this works as well.)

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java (hunk 6 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java]: done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java (hunk 7 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CombineFnTesterTest.java]: I have done both.

**Pull request closed without merging**